### PR TITLE
Improve env for CPU benchmarking

### LINKED ISSUE #37: Improve CPU benchmarking remote environment setup
CPU benchmarking with ONNXRuntime does not have any pre-built containers. So we have to rely on using a conda environment with the latest versions of the libraries.

### PR DESCRIPTION
**Pull request by ramkrishna2910**
This PR adds an automated way of creating a conda environment for benchmarking on CPUs using OnnxRuntime.
Closes #37

### CODE REVIEW

**Comment by jeremyfowers** [comment]
Just to be clear @ramkrishna2910:
- this PR is cleaning up the code 
- but the user shouldn't expect to see any change to results, how long it takes to run a benchmark, etc.?

**Review by jeremyfowers — CHANGES_REQUESTED** [review]
CPU benchmarking no longer works for me:

```
(mla) jfowers@jfowers:~/mlagility$ python examples/api/hello_world.py 
2023-02-02 08:16:17.230248: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.

Woohoo! Build "hello_world" (build_name auto-selected) found in cache. Loading it!

Info: GPU is not used for accuracy comparisons it's only used for performance comparison. So inputs provided during model compilation is used.
      User is responsible for ensuring the remote GPU server is turned on and has python>=3.8, docker>=20.10 installed.

Performance of build hello_world on device nvidia is:
latency: 0.034 ms
throughput: 21685.2 ips

Woohoo! Build "hello_world" (build_name auto-selected) found in cache. Loading it!

Info: CPU is not used for accuracy comparisons it's only used for performance comparison. So dummy inputs are used.
      User is responsible for ensuring the remote cpu server is turned on and has python>=3.8 installed.
Traceback (most recent call last):
  File "examples/api/hello_world.py", line 24, in <module>
    benchit(pytorch_model, inputs, device="x86")
  File "/home/jfowers/mlagility/src/mlagility/api.py", line 145, in benchit
    perf = cpu_model.benchmark()
  File "/home/jfowers/mlagility/src/mlagility/benchmark/cpumodel.py", line 56, in benchmark
    benchmark_results = self._execute(repetitions=repetitions)
  File "/home/jfowers/mlagility/src/mlagility/benchmark/cpumodel.py", line 83, in _execute
    cloud.execute_cpu_remotely(self.state, self.log_execute_path, repetitions)
  File "/home/jfowers/mlagility/src/mlagility/benchmark/cloud.py", line 311, in execute_cpu_remotely
    s.get(
  File "/net/home/jfowers/miniconda3/envs/mla/lib/python3.8/site-packages/paramiko/sftp_client.py", line 811, in get
    size = self.getfo(remotepath, fl, callback, prefetch)
  File "/net/home/jfowers/miniconda3/envs/mla/lib/python3.8/site-packages/paramiko/sftp_client.py", line 782, in getfo
    file_size = self.stat(remotepath).st_size
  File "/net/home/jfowers/miniconda3/envs/mla/lib/python3.8/site-packages/paramiko/sftp_client.py", line 493, in stat
    t, msg = self._request(CMD_STAT, path)
  File "/net/home/jfowers/miniconda3/envs/mla/lib/python3.8/site-packages/paramiko/sftp_client.py", line 822, in _request
    return self._read_response(num)
  File "/net/home/jfowers/miniconda3/envs/mla/lib/python3.8/site-packages/paramiko/sftp_client.py", line 874, in _read_response
    self._convert_status(msg)
  File "/net/home/jfowers/miniconda3/envs/mla/lib/python3.8/site-packages/paramiko/sftp_client.py", line 903, in _convert_status
    raise IOError(errno.ENOENT, text)
FileNotFoundError: [Errno 2] No such file
```

File: src/mlagility/benchmark/cloud.py [hunk 1]
```diff
@@ -283,19 +284,18 @@ def execute_cpu_remotely(
...
+    process = subprocess.Popen(["bash", f"mlagility_remote_cache/setup_ort_env.sh {env_name}"],\
+                                stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    process.communicate()
 
+    print("Running benchmarking script...")
     _, exit_code = exec_command(
         client,
         (
-            "/usr/bin/python3 mlagility_remote_cache/execute-cpu.py "
+            f"/home/{username}/miniconda3/envs/{env_name}/bin/python mlagility_remote_cache/ \
```
Comment (jeremyfowers) [inline src/mlagility/benchmark/cloud.py:297]:
```suggestion
            f"/home/{username}/miniconda3/envs/{env_name}/bin/python mlagility_remote_cache/"
            "execute-cpu.py"
```

**Comment by ramkrishna2910** [comment]
> 



> Just to be clear @ramkrishna2910:
> 
> * this PR is cleaning up the code
> * but the user shouldn't expect to see any change to results, how long it takes to run a benchmark, etc.?

Yes that is correct.

**Comment by ramkrishna2910** [comment]
> CPU benchmarking no longer works for me:
> 
> ```
> (mla) jfowers@jfowers:~/mlagility$ python examples/api/hello_world.py 
> 2023-02-02 08:16:17.230248: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
> To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
> 
> Woohoo! Build "hello_world" (build_name auto-selected) found in cache. Loading it!
> 
> Info: GPU is not used for accuracy comparisons it's only used for performance comparison. So inputs provided during model compilation is used.
>       User is responsible for ensuring the remote GPU server is turned on and has python>=3.8, docker>=20.10 installed.
> 
> Performance of build hello_world on device nvidia is:
> latency: 0.034 ms
> throughput: 21685.2 ips
> 
> Woohoo! Build "hello_world" (build_name auto-selected) found in cache. Loading it!
> 
> Info: CPU is not used for accuracy comparisons it's only used for performance comparison. So dummy inputs are used.
>       User is responsible for ensuring the remote cpu server is turned on and has python>=3.8 installed.
> Traceback (most recent call last):
>   File "examples/api/hello_world.py", line 24, in <module>
>     benchit(pytorch_model, inputs, device="x86")
>   File "/home/jfowers/mlagility/src/mlagility/api.py", line 145, in benchit
>     perf = cpu_model.benchmark()
>   File "/home/jfowers/mlagility/src/mlagility/benchmark/cpumodel.py", line 56, in benchmark
>     benchmark_results = self._execute(repetitions=repetitions)
>   File "/home/jfowers/mlagility/src/mlagility/benchmark/cpumodel.py", line 83, in _execute
>     cloud.execute_cpu_remotely(self.state, self.log_execute_path, repetitions)
>   File "/home/jfowers/mlagility/src/mlagility/benchmark/cloud.py", line 311, in execute_cpu_remotely
>     s.get(
>   File "/net/home/jfowers/miniconda3/envs/mla/lib/python3.8/site-packages/paramiko/sftp_client.py", line 811, in get
>     size = self.getfo(remotepath, fl, callback, prefetch)
>   File "/net/home/jfowers/miniconda3/envs/mla/lib/python3.8/site-packages/paramiko/sftp_client.py", line 782, in getfo
>     file_size = self.stat(remotepath).st_size
>   File "/net/home/jfowers/miniconda3/envs/mla/lib/python3.8/site-packages/paramiko/sftp_client.py", line 493, in stat
>     t, msg = self._request(CMD_STAT, path)
>   File "/net/home/jfowers/miniconda3/envs/mla/lib/python3.8/site-packages/paramiko/sftp_client.py", line 822, in _request
>     return self._read_response(num)
>   File "/net/home/jfowers/miniconda3/envs/mla/lib/python3.8/site-packages/paramiko/sftp_client.py", line 874, in _read_response
>     self._convert_status(msg)
>   File "/net/home/jfowers/miniconda3/envs/mla/lib/python3.8/site-packages/paramiko/sftp_client.py", line 903, in _convert_status
>     raise IOError(errno.ENOENT, text)
> FileNotFoundError: [Errno 2] No such file
> ```

Fixed

**Review by jeremyfowers — APPROVED** [review]
It failed for me on the first try but worked on the second and third tries. Lets move on for now, but can you file a bug for figuring out what goes wrong in a fresh conda env?

**Comment by ramkrishna2910** [comment]
> It failed for me on the first try but worked on the second and third tries. Lets move on for now, but can you file a bug for figuring out what goes wrong in a fresh conda env?

Filed #64  for later investigation

**Pull request merged**